Add the Kyup container provider API to the cloud modules

hackman opened the pull request:
##### ISSUE TYPE
- New Module Pull Request
##### COMPONENT NAME

kyup
##### ANSIBLE VERSION

```
ansible 2.2.0
  config file = /root/api/ansible.cfg
  configured module search path = Default w/o overrides
```
##### SUMMARY

Introducing the Kyup container provider API to Ansible.
With this module, users of Ansible will be able to provision and manage containers inside the Kyup architecture.
The module has been developed and tested on Debian Jessie.

hackman committed:
Initial release of the KYUP API Ansible module

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Check for KYUP_API_KEY only in the core()
- Removed redundant api_key check and replaced them with direct call of module.params.get('api_key')
- Small rearangement of the DOCUMENTATION section. Updated the version_added.

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: removed unused getkeyordie()

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: fixed a few integer casting issues
- kyup_action() made a cast to integer for container_id
- get_task_status()
  - made a cast to integer for task_id
  - made the error reporting more robust
- create_container(), moved the whole request formating in the actual request, so it will not be separate and this way it will be more apperant, what we are doing

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Optimized the code to use a single function for API requests
- Moved all the request and "status" handling code into a single function api_request()
- Rewrote kyup_action and create_container to use the new function
- Moved the retries count next to the API url definition, because this way it is easily visible as a configuration variable instead of a hardcoded value
- Moved the URL below the imports

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: introduced functionality to add ssh keys
- new function add_ssh_keys() added
  If no ssh_keys parameter is supplied in the playbook, it returns immediately
  If a comma separated list is supplied it will issue sshInstallKey call to the API for each key supplied
  In order to add the keys it will first list all kays and find their IDs
  Also any spaces in the ssh_keys parameter will be removed

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Replaced all double quotes with single quotes to make the code more uniform

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Introduced ansible_facts
- Replaced all double quotes with single quotes to make the code more uniform

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Check if a container with the same name already exists

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Removed the container_id requirement as most people will want to use names instead
- removed the mutually_exclusive list, as it is no longer relevant
- removed container_id parameter from almost all calls to kyup_action()
- moved the container_id parameter as the last parameter of kuyp_action() and given it a default value. This also removed the need to initialize container_id later in the function
- added code to get the container_id from the name supplied in the playbook by calling the API
- in create container added ansible_facts
- reintroduced the error handling code in api_request()
- removed the container_id parameter as a whole, also removed the documentation for it

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Updated the examples with tested playbook examples

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
Merge branch 'devel' of github.com:Kyup-com/ansible-modules-core into devel

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Updated documentation
- Added required option 'name'
- removed the default options in the documentation, as we don't actually set null as a value

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Removed the aliases and choices from the DOCUMENTATION
- Removed aliases and choices as the http://docs.ansible.com/ansible/developing_modules.html#module-checklist requires
- Set the default dc_id to USA

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Added more requirements to the documentation of parameters password, image and datacenter

Signed-off-by: Marian Marinov <mm@yuhu.biz>

gregdek commented:
Thanks @hackman for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

hackman committed:
cloud/kyup/kyup.py: Replaced tabs with spaces in the DOCUMENTATION

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Better handling of os.environ['KYUP_API_KEY']

Signed-off-by: Marian Marinov <mm@yuhu.biz>

cnicolov commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,338 @@
+    description:
+      - Template image to use when creating the container. Required when I(action) = 'create'.
+    required: false
+  dc_id:
```
You can add `aliases:` for better documentation representation.

hackman replied:
I found somewhere in the documentation that aliases should be removed from DOCUMENTATION and that is why I removed them

cnicolov replied:
I think this is valid only for the empty `aliases: []` and `choices: []` statements but I'm not sure

hackman replied:
commit 63597f8f78c8db2fd8b9007654e9d171264a05f3
removes aliases and choices as http://docs.ansible.com/ansible/developing_modules.html#module-checklist requires.
The checklist is not specific if this rule is intended only for empty lists. But I prefer to be safe. So I removed them.

ryansb replied:
That's only for empty ones in docs. It's not very nice to read (as a user) a bunch of empty `aliases:` lists when looking for something. Aliases are still allowed.

cnicolov commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,338 @@
+'''
+
+import os
+import requests
```
If executed in a python environment without requests installed the module will fail.
For example when this task needs to be delegated to a host with internet access(NAT) in a controlled environment.

hackman replied:
It took me some time to replace requests with fetch_url from the urls.py, but I did it :)

cnicolov commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,338 @@
+
+def api_request(module, data):
+    ret = {}
+    count = retries;
```
semicolon token is required only when you're calling more than one statement on the same line

hackman replied:
yup... a habit from other languages. Fixed

cnicolov commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,338 @@
+def api_request(module, data):
+    ret = {}
+    count = retries;
+    while count != 0:
```
`while count:` will do the trick here

hackman replied:
I do prefer to actually write it this way... but if you require it like that. I'll change it.

cnicolov commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,338 @@
+        count -= 1
+        time.sleep(5)
+
+    if 'status' in ret:
```
You can do 

``` python
status = ret.get('status', False)
```

and use

``` python
if not status:
```

and get rid of the `else` statements?

hackman replied:
I did what you requested... however I don't like it this way :(

cnicolov commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,338 @@
+def get_task_status(module, task_id):
+    req = '{"action":"getTask","authorization_key":"%s","data":{"task_id":%d}}'
+    count = retries;
+    while count != 0:
```
`while count` will do the trick

hackman replied:
I do prefer to actually write it this way... but if you require it like that. I'll change it.

cnicolov commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,338 @@
+    if ssh_keys is None or ssh_keys == '':
+        return
+    # cleanup any spaces put by the user
+    ssh_keys.replace(' ', '')
```
IMO list is better fit for the `ssh_keys` argument.

The syntax will become

``` yaml
kyup:
  ssh_keys:
    - foo
    - bar

kyup: >
  ssh_keys=["foo", "bar"]
```

hackman replied:
I'm sorry but I do not agree. List is harder to be written in the yml file. And I prefer the comma separated list

cnicolov replied:
I was thinking for something like this: 

``` yaml
# group_vars/prod_web/keys.yml
devops_organization_pubkeys:
  - name: peter
    pubkey: ssh-blabla really long string of nonsense peter@devops.organization
  - name: bob
    pubkey: ssh-blalbla really long string of nonsense bob@devops.organization
```

If you decide to extend the module with a feature which allows the users to create new keys:

``` yaml
- name: "Add ssh pubkeys"
  kyup:
    action: ssh_key
    name: "{{ item.name }}"
    pubkey: "{{ item.pubkey }}"
  with_items:
    - "{{ devops_organization_pubkeys }}"
```

Now let's create new container:

``` yaml
- name: "Create new container"
  kyup:
    action: create
    ssh_key: "{{ devops_organization_keys|map(attribute='name')|list }}"
```

This also can be done with comma separated string:

``` yaml
"{{ devops_organization_keys|map(attribute='name')|join(",") }}"
```

Now when the user needs a new pubkey he/she can add it as a group var.

Anyway..

IMO if something can get multiple values as input it should be a collection. It is cleaner and easier to read/follow. And IMO it's not that important how hard it was to write the thing once but how you'll read it for example 500 times.

Consider the following scenario:
 There is a typo in the code - a comma in the middle of the string `peter`:

``` yaml
kyup:
  ssh_keys: "bob,pe,ter,alice"
```

The error might look like:

```
Error! SSH key pe does not exist
```

Example with a list:

``` yaml
kyup:
  ssh_keys:
    - "bob"
    - "pe,ter"
    - "alice"
```

```
Error! SSH key pe,ter does not exist
```

I think the second case is cleaner.

cnicolov commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,338 @@
+        api_request(module, req % ('sshInstallKey', module.params.get('api_key'), key_req % (int(key_list[key]), int(container_id) )))
+
+def create_container(module):
+    enc_key = module.params['enc_key'] or os.environ['KYUP_ENC_KEY']
```
If `module.params['enc_key']` is false and `KYUP_ENC_KEY` env var isn't set this will raise an KeyError exception.

`os.environ.get('KYUP_ENC_KEY', None)` will do the trick

hackman replied:
Yup, I saw that one, also :)

cnicolov commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,338 @@
+            container['ip'] = container['ip'][0:-3]
+            module.exit_json(
+                changed=False,
+                ansible_facts = { 'container_name': container['name'], 'container_ip': container['ip'] },
```
You can prefix the facts with the module name to avoid overrides

hackman replied:
done

cnicolov commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,338 @@
+    elif action == 'create':
+        create_container(module)
+    else:
+        if action == 'destroy':
```
You can create a class with static method for each action and call them with getattr, this might simplify the extensibility for the `action` argument. for example

``` python
class Actions:
    @staticmethod
    def create(module):
        return create_container(module)

action = getattr(Actions, action)
action(module)
```

hackman replied:
for now, if this is not required I would not do that.

ryansb replied:
That sort of thing (using getattr to avoid if/else checks) isn't something we usually ask for as part of modules, since it's primarily a question of style.

cnicolov commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,338 @@
+            api_key = dict(aliases=['API_KEY'], no_log=True),
+            enc_key = dict(aliases=['ENC_KEY'], no_log=True),
+            action = dict(type='str', aliases=['state', 'command', 'cmd']),
+            dc_id = dict(type='int', aliases=['datacenter', 'datacenter_id'], default=1),
```
I think `dc` is also an suitable alias for this argument

hackman replied:
agreed

cnicolov reviewed (commented):
Also you can look at [PEP8](https://www.python.org/dev/peps/pep-0008/) to make the code more readable.

This is only an opinion, so feel free to ignore my comments

hackman committed:
cloud/kyup/kyup.py: Replaced the requests module with fetch_url and removed json in favour of the functions from basic.py
- removed the import of requests
- removed the import of json
- added ansible.module_utils.urls in order to get fetch_url
- changed the api_request function to use both fetch_url and module.from_json()

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Updated the version_added to 2.3

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Added dc and dc_id aliases for the datacenter option

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Update the api_key handling, once more. This time use os.environ.get() for proper handling of environment variables

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Prefixed the facts with 'kyup_' in order to make them unique

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Fixed a situation where enc_key may throw an exeption

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Style changes for the while loops

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: renamed variable status to http_status in api_request(), because otherwise it is colliding with the status from within the response of the API

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Logic simplification in api_request()
- Changed the way I get the status field from the response with .get() and now I may be sure that the var is always set.
 The new approach simplified the logic a bit.

Signed-off-by: Marian Marinov <mm@yuhu.biz>

hackman committed:
cloud/kyup/kyup.py: Added RETURN documentation

Signed-off-by: Marian Marinov <mm@yuhu.biz>

cnicolov commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,357 @@
+        module.fail_json(changed=False, msg = 'Req: ' + data + ' Resp: ' + ret)
+
+    # We have found status, check its value here
+    if status:
```
You can omit the `if/else` statement and just return here.

The else statement won't get executed, the module will exit at line 188 if the status is `False`

hackman replied:
I do not agree. The status variable can be any of the following values:
0, 1, 101, 102 and so on.
The first check is if the status variable has assigned value.
The second check is if the value is true, if not do something else.

cnicolov replied:
```
>>> var.update({"status": 0})
>>> if var.get("status", False):
...     print "wont print"
...
>>> var.get("status")
0
>>> var.get("status") == False
True
```

If you don't like this behavior you must use `if status != 0:` to make difference if `status` is `integer` or even better, use the get method with `None` as fallback.

```
>>> var.get('doesnotexist', None) is None
True
```

cnicolov commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,357 @@
+        module.fail_json(changed=True, msg = 'Unable to get task_id')
+
+def core(module):
+    api_key = module.params['api_key']
```
You can do:

``` python
api_key = os.environ.get('KYUP_API_KEY', None) or module.params['api_key']

if not api_key:
     module.fail_json(changed=False, msg='you can not continue without api_key')
```

and omit the two `if` statements below

hackman replied:
I want to:
1. get the api_key form the playbook
2. if env variable is existing overwrite the api_key
3. check if the api_key is not None
As far as my tests go, the code does exactly what I need.

cnicolov commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,357 @@
+    if enc_key is None:
+        module.fail_json(changed=False, msg = 'ENC_KEY is required for creation of containers')
+
+    opt = { 'name' : None, 'password' : None, 'image' : None, 'dc_id' : None }
```
If more than one of the parameters is None, you can detect it in one shot for better user experience:

``` python
mandatory_parameters = ["name', 'password', 'image', 'dc_id']
missing = [p for p in mandatory_parameters if module.params.get(p, None) is None]
if len(missing):
    module.fail_json(
        changed=False,
        msg='One or more parameters are missing: %s' % missing.join(","))
```

gregdek commented:
@hackman A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

hackman commented:
All issues have been resolved.
ready_for_review

gregdek commented:
Thanks @hackman for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

hackman commented:
Is there any news for this?

hackman commented:
Ping

ryansb commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,357 @@
+      dc_id=1
+      name=avalon
+      password=e4a17265b28f6a9e8657
+      ssh_keys=hackman,piele
```
As mentioned in a later comment, a `list` type is preferred for this parameter since it is actually a list of things.

ryansb commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,357 @@
+      - Number of CPU cores to be used. Minimum 2 CPU cores.
+    required: false
+    default: 2
+  bw:
```
Super short names like this aren't obvious to everyone, can you name this something like "bandwidth" and use an alias to `bw` for power users?

ryansb commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,357 @@
+      - DataCenter location ID. Required when I(action) = 'create'.
+    required: false
+    default: 1
+  mem:
```
I'd prefer using `memory` for this parameter to be more consistent with other modules. Also, is a decimal value ok for this field? For example, does kyup allow 1.5GB machines?

ryansb commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,357 @@
+    padding = blocksize - ( len(data) % blocksize )
+    return data + (chr(padding) * padding)
+
+def encrypt( key, data ):
```
Can you please remove the whitespace around these params to fit with PEP8 code style?

ryansb commented on cloud/kyup/kyup.py:
```diff
@@ -0,0 +1,357 @@
+        storage_type = 1
+
+    container_id = 0
+    req = '{"action":"cloudCreate","authorization_key":"%s","data":{"name":"%s","password":"%s","image_name":"%s","datacenter_id":%d,"storage_type":%d,"resources":{"mem":%d,"hdd":%d,"cpu":%d,"bw":%d}}}'
```
Please use the JSON serializer instead. Using string formatting to inject things into JSON could have unfortunate consequences.
